### ISSUE TITLE
some randomness in first 30 moves

### ISSUE THREAD

**Issue by petgo3** [gcp/leela-zero#763]
Excuse me for my rather simple question:
What kind of some randomness uses the training?
"Training games are played with some randomness in first 30 moves"
Just completely random by let's say 5% probability or are there more specific rules?

**Comment by jkiliani** [comment]
No, for the first 30 moves the tree search result is used proportionally, i.e. if the search is 60% of rollouts move A, 30% move B and 10% move C, there's a 30% chance that move B is played and a 10% chance for move C, even though the search found move A objectively better. From move 31 on, in the above example, move A would always be picked (this is called greedy selection).

**Comment by petgo3** [comment]
Thank you for the explanation.  
Then it seems, that one opponent could be lucky (to get the best moves) while the other suffers due to this randomness (="bad" moves). This leads to noise in winning statistics and the quality of games is lower.
Is there any method to distribute the "bad" moves equaly to both colors. e.g. reduce the proportionally by the amount of allready played "bad" moves of one color or the sum of probability of missed best moves ...
This might help to get more even starting positions from move 30 on.

**Comment by panda0307** [comment]
In AGZ paper, why the temperature->0 means that selects a move with maximum visit count?
I don't understand it from the formula: π(a|s)=N(s, a)^1/τ / ΣN(s , b)^1/τ

**Comment by Dorus** [comment]
> Then it seems, that one opponent could be lucky (to get the best moves) while the other suffers due to this randomness (="bad" moves). This leads to noise in winning statistics and the quality of games is lower.

Actually both players have that same problem so the odds the player that is already ahead wins is still larger. Because this player need to make more mistakes. Also it can only make mistakes if it searched these mistakes. A stronger net wouldn't even search them.

Also the policy output is trained on the predicted and not the selected moves, thus the training data does not suffer from the more random move prediction.

The value net does get affected as the winner is decided by the selected moves. However as mentioned above the winner will still be correct for most moves (and only the first 30 are more random, so the other 250 moves are still stabel). 

On top of that, AZ used proportionally move selection all game, and this seemed to work too (they used t=1).

The theory is that a slightly more random value output does not offset the gain from having widely more variation in game output. Without temperature, almost all games would be identical and it would take a long time before it starts evaluating move sequences it consider slightly less favourable.

Also if a move pops up in the policy net, with temperature it has a higher chance to be selected and future evaluated so that the next net can learn more about this move, as follow up moves get trained and become higher quality, thus the next time it is searched, the search is higher quality and will decide if the move is good or bad quicker. A bad move will then get less playouts and a good move more, so the proportionally move selection will then pick the good move even more and the bad move less, even if the good move did not yet increase enough to be selected as best.

On AZ they did not run evaluation so it might even have been important to run t=1 in order to prevent over fitting. Instead they could afford to train with lower learning rate and move toward the game data slower. If a bad move was selected because of bad training, the next set of selfplay games would contain more data on this move and as the analysis on this move improves, it would fall off again.

**Comment by jkiliani** [comment]
@panda0307 The formula has to be understood in the way that for t->0, the exponent applied to the move counts approaches infinity. This makes the propbability of the move with the largest playout count infinitely larger than that for the move with even the next largest count, and even in the case when that is only slightly smaller.

My personal opinion is that temperatures between 1/4 and 1/2 might very well be worth trying, since they allow moves with slightly less visits to be selected on occasion, while already effectively preventing moves with very low visit count. However, that is currently not implemented in Leela Zero, and I doubt it will be either.

**Comment by holicst** [comment]
May I ask why 30 moves are used? Can we experiment with lowering and raising this and see if progress is affected?

**Comment by jkiliani** [comment]
I think 30 moves were used in order to have reasonably diverse openings, which would already provide diversity in the patterns encountered in training games, while affecting the strength of the value net as little as possible. But you're right, @gcp has also called this a "magic number", which is a less than ringing endorsement to the concept.

**Comment by holicst** [comment]
Perhaps having this random window in the midgame or endgame can help enhancing those specific areas? Possibly a sliding N-move randomness window could help. Also, I think this may have different optimal settings when the training is at the very beginning and when the network is already on the dan level.

**Comment by petgo3** [comment]
My point was more the balance of "bad" or experimental moves and game quality.
If for example only black plays 3 experimental moves, white will win perhaps even with a weaker network.
This leads to an unbalanced game and might produce less information to learn.

So it might be interesting to keep the balance on experimentel moves somehow.

**Comment by Dorus** [comment]
But match games are played with optimal strength and no additional randomness or noise. Only selfplay games have that.

**Comment by Dorus** [comment]
But on self-play, the network plays against itself. So 

> white will win perhaps even with a weaker network.

makes no sense.
